test/e2e/common.go

```diff
@@ -212,3 +214,43 @@ func createRestConfig(ctx context.Context, tmpdir, namespace, clusterName string
 
 	return config
 }
+
+// EnsureControlPlaneInitialized waits for the cluster KubeadmControlPlane object to be initialized
+// and then installs cloud-provider-azure components via helm
+// Fulfills the clusterctl.Waiter type so that it can be used as ApplyClusterTemplateAndWaitInput data
+// in the flow of a clusterctl.ApplyClusterTemplateAndWait E2E test scenario
+func EnsureControlPlaneInitialized(ctx context.Context, input clusterctl.ApplyClusterTemplateAndWaitInput, result *clusterctl.ApplyClusterTemplateAndWaitResult) {
+	getter := input.ClusterProxy.GetClient()
+	cluster := framework.GetClusterByName(ctx, framework.GetClusterByNameInput{
+		Getter:    getter,
+		Name:      input.ConfigCluster.ClusterName,
+		Namespace: input.ConfigCluster.Namespace,
+	})
+	kubeadmControlPlane := &kubeadmv1.KubeadmControlPlane{}
+	key := crclient.ObjectKey{
+		Namespace: cluster.Spec.ControlPlaneRef.Namespace,
+		Name:      cluster.Spec.ControlPlaneRef.Name,
+	}
+	Eventually(func() error {
+		return getter.Get(ctx, key, kubeadmControlPlane)
+	}, input.WaitForControlPlaneIntervals...).Should(Succeed(), "Failed to get KubeadmControlPlane object %s/%s", cluster.Spec.ControlPlaneRef.Namespace, cluster.Spec.ControlPlaneRef.Name)
+	if kubeadmControlPlane.Spec.KubeadmConfigSpec.ClusterConfiguration.ControllerManager.ExtraArgs["cluster-cidr"] == "10.244.0.0/16,2001:1234:5678:9a40::/58" {
+		InstallCalicoHelmChartForDualStack(ctx, input)
+	} else if kubeadmControlPlane.Spec.KubeadmConfigSpec.ClusterConfiguration.ControllerManager.ExtraArgs["cluster-cidr"] == "2001:1234:5678:9a40::/58" {
```

Comment:
Also @aramase is the equivalent check here to check for an array of len == 1, and to detect for a string that "is an ipv6 cidr"? (hopefully there's a standard library I can use for that, I'll look around)

Reply:
> Also @aramase is the equivalent check here to check for an array of len == 1, and to detect for a string that "is an ipv6 cidr"? (hopefully there's a standard library I can use for that, I'll look around)

Yeah, checking the CIDR family is the right way. You can use the [net](https://pkg.go.dev/k8s.io/utils@v0.0.0-20220210201930-3a6ce19ff2f9/net) package in `k8s.io/utils` to check if it's [IPv6CIDRString](https://pkg.go.dev/k8s.io/utils@v0.0.0-20220210201930-3a6ce19ff2f9/net#IsIPv4CIDRString).